stages/org.osbuild.coreos.live-artifacts.mono

```diff
@@ -0,0 +1,937 @@
+        "iso/images/pxeboot/vmlinuz": "",
+        "iso/isolinux": "",
+        "iso/zipl.prm": "",  # s390x only, deleted after use
+        "initrd": "",
+        "initrd/etc/coreos/features.json": "",
+        "initrd/etc/coreos-live-want-rootfs": "",
+        "initrd/etc/coreos-live-initramfs": "",
+        "initrd-rootfs": "",
+        "initrd-rootfs/root.squashfs": "",
+    }
+    for key in paths:
+        paths[key] = os.path.join(workdir, key)
+    return paths
+
+
+def mk_workdirs(paths):
+    for d in (paths["iso"],
+              paths["iso/coreos"],
+              paths["iso/isolinux"],
+              paths["iso/images"],
+              paths["iso/images/pxeboot"],
+              paths["initrd"],
+              paths["initrd-rootfs"]):
+        os.mkdir(d)
+
+
+def main(workdir, tree, inputs, options, loop_client):
+    """
+    Here we will generate the 4 Live/PXE CoreOS artifacts.
+    There are four files created and exported from this stage.
+
+    1. live-iso
+    2. live-kernel    (i.e. images/pxeboot/vmlinuz)
+    3. live-initramfs (i.e. images/pxeboot/initrd.img)
+    4. live-rootfs    (i.e. images/pxeboot/rootfs.img)
+
+    The live-iso itself contains the other 3 artifacts inside of
+    it. A rough approximation of the structure looks like:
+
+      live-iso.iso
+        -> images/pxeboot/vmlinuz
+        -> images/pxeboot/initrd.img
+        -> images/pxeboot/rootfs.img
+          -> metal.osmet
+          -> metal4k.osmet
+          -> root.squashfs
+            -> full filesystem of metal image
+
+    Where the ISO contains the kernel and initrd and the rootfs. The
+    rootfs contains the osmet file and the root.squashfs, which itself
+    contains all the files from a full CoreOS system.
+    """
+
+    img_metal, img_metal4k = parse_metal_inputs(inputs)
+    # The deployed tree input is a deployment tree just as it would
+    # appear on a booted OSTree system. We copy some files out of this
+    # input tree since it is easier and they match what is in the metal
+    # images anyway. We also use it as a chroot when we run `coreos-installer`.
+    deployed_tree = inputs['deployed-tree']['path']
+
+    # Determine some basic information about the CoreOS we are operating on.
+    os_release = osrelease.parse_files(os.path.join(deployed_tree, 'etc', 'os-release'))
+    version = os_release['OSTREE_VERSION']
+
+    paths = mk_paths(workdir)
+    mk_workdirs(paths)
+
+    # Find the directory under `/usr/lib/modules/<kver>` where the
+    # kernel/initrd live. It will be the only entity in there.
+    modules_dir = os.path.join(deployed_tree, 'usr/lib/modules')
+    modules_dirents = os.listdir(modules_dir)
+    if len(modules_dirents) != 1:
+        raise ValueError(f"expected unique entry in modules dir, found: {modules_dirents}")
+    moduledir = modules_dirents[0]
+
+    # copy those files out of the ostree into the iso root dir
+    for file in ['vmlinuz', 'initramfs.img']:
```

Comment:
I'll unroll it.

I'm not sure about the chmod.. As you mentioned, I'm seeing on my local FCOS system:

```
core@apu2:~$ ls -lh /usr/lib/modules/6.11.8-300.fc41.x86_64/initramfs.img 
-rw-r--r--. 1 root root 91M Jan  1  1970 /usr/lib/modules/6.11.8-300.fc41.x86_64/initramfs.img
core@apu2:~$ ls -lh /usr/lib/modules/6.11.8-300.fc41.x86_64/vmlinuz 
-rwxr-xr-x. 1 root root 16M Jan  1  1970 /usr/lib/modules/6.11.8-300.fc41.x86_64/vmlinuz
```

@jlebon should we drop the `chmod`?

Reply:
Indeed, seems like the initramfs is already 0644 in FCOS and RHCOS at least, so we could drop it I guess. That said, I don't think we have any tests that would fail today if it reverted in the OS. It would make sense I think to add a check in our of our tests that it remains 0644. This originates from https://github.com/coreos/coreos-assembler/commit/3f349e39eb3778236d99fc7c6f7ccfe1f7d2f868 which presents a reasonable use case.

Reply:
Will unroll and drop the chmod in the next upload. @jlebon - I'll add a test to make sure the permissions of the initrams.img doesn't change.